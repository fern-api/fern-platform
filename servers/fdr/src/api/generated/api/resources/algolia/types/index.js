export * from "./AlgoliaSearchIndex";
export * from "./Breadcrumbs";
export * from "./IndexSegmentId";
export * from "./AlgoliaRecord";
export * from "./AlgoliaRecordVersionV3";
export * from "./AlgoliaPageRecordV3";
export * from "./AlgoliaEndpointRecordV3";
export * from "./AlgoliaWebSocketRecordV3";
export * from "./AlgoliaWebhookRecordV3";
export * from "./IndexSegment";
export * from "./SingleAlgoliaIndexInfo";
export * from "./UnversionedSingleAlgoliaIndexInfo";
export * from "./VersionedSingleAlgoliaIndexInfo";
export * from "./SearchInfo";
export * from "./LegacyMultiAlgoliaIndexInfo";
export * from "./AlgoliaEndpointRecord";
export * from "./AlgoliaPageRecord";
export * from "./AlgoliaRecordEndpointPath";
export * from "./AlgoliaRecordPathPart";
export * from "./AlgoliaRecordPath";
export * from "./AlgoliaRecordEndpointSummary";
export * from "./AlgoliaRecordVersion";
export * from "./AlgoliaEndpointRecordV2";
export * from "./AlgoliaPageRecordV2";
