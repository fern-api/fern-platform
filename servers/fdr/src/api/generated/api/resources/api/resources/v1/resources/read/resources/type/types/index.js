export * from "./TypeId";
export * from "./TypeDefinition";
export * from "./TypeShape";
export * from "./TypeReference";
export * from "./ObjectType";
export * from "./PropertyKey";
export * from "./ObjectProperty";
export * from "./PrimitiveType";
export * from "./IntegerType";
export * from "./DoubleType";
export * from "./StringType";
export * from "./OptionalType";
export * from "./ListType";
export * from "./SetType";
export * from "./MapType";
export * from "./EnumType";
export * from "./EnumValue";
export * from "./UndiscriminatedUnionType";
export * from "./UndiscriminatedUnionVariant";
export * from "./DiscriminatedUnionType";
export * from "./DiscriminatedUnionVariant";
export * from "./LiteralType";
export * from "./FormDataRequest";
export * from "./BytesRequest";
export * from "./FormDataProperty";
export * from "./FormDataBodyProperty";
export * from "./FormDataFileProperty";
export * from "./FilePropertySingle";
export * from "./FilePropertyArray";
export * from "./ContentType";
