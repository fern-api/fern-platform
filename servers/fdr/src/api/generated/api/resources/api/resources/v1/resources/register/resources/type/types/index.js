"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./TypeId"), exports);
__exportStar(require("./TypeDefinition"), exports);
__exportStar(require("./TypeShape"), exports);
__exportStar(require("./TypeReference"), exports);
__exportStar(require("./ObjectType"), exports);
__exportStar(require("./PropertyKey"), exports);
__exportStar(require("./ObjectProperty"), exports);
__exportStar(require("./PrimitiveType"), exports);
__exportStar(require("./OptionalType"), exports);
__exportStar(require("./ListType"), exports);
__exportStar(require("./SetType"), exports);
__exportStar(require("./MapType"), exports);
__exportStar(require("./EnumType"), exports);
__exportStar(require("./EnumValue"), exports);
__exportStar(require("./UndiscriminatedUnionType"), exports);
__exportStar(require("./UndiscriminatedUnionVariant"), exports);
__exportStar(require("./DiscriminatedUnionType"), exports);
__exportStar(require("./DiscriminatedUnionVariant"), exports);
__exportStar(require("./LiteralType"), exports);
__exportStar(require("./FileUploadRequest"), exports);
__exportStar(require("./BytesRequest"), exports);
__exportStar(require("./FileUploadRequestProperty"), exports);
__exportStar(require("./FileProperty"), exports);
__exportStar(require("./FilePropertySingle"), exports);
__exportStar(require("./FilePropertyArray"), exports);
