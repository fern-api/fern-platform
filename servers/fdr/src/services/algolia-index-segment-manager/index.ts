export {
    AlgoliaIndexSegmentManagerServiceImpl,
    type AlgoliaIndexSegmentManagerService,
} from "./AlgoliaIndexSegmentManagerService";
