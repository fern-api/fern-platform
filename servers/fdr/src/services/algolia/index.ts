export { AlgoliaServiceImpl, type AlgoliaService } from "./AlgoliaService";
export type { AlgoliaSearchRecord, ConfigSegmentTuple, IndexSegment } from "./types";
