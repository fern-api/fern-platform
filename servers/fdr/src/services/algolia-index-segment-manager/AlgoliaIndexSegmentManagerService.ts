import { visitDbNavigationConfig } from "@fern-api/fdr-sdk";
import { addHours, addMinutes } from "date-fns";
import { kebabCase } from "lodash-es";
import { v4 as uuidv4 } from "uuid";
import { Cache } from "../../Cache";
import { DocsV1Db } from "../../api";
import type { FdrApplication } from "../../app";
import type { DocsVersion } from "../../types";
import type { ConfigSegmentTuple, IndexSegment } from "../algolia";

const SECONDS_IN_ONE_HOUR = 60 * 60;

type GenerateNewIndexSegmentsResult =
    | {
          type: "versioned";
          configSegmentTuples: ConfigSegmentTuple[];
      }
    | {
          type: "unversioned";
          configSegmentTuple: ConfigSegmentTuple;
      };

export interface AlgoliaIndexSegmentManagerService {
    getOrGenerateSearchApiKeyForIndexSegment(indexSegmentId: string): string;

    getSearchApiKeyForIndexSegment(indexSegmentId: string): string | undefined;

    generateAndCacheApiKey(indexSegmentId: string): string;

    generateIndexSegmentsForDefinition({
        dbDocsDefinition,
        url,
    }: {
        dbDocsDefinition: DocsV1Db.DocsDefinitionDb;
        url: string;
    }): GenerateNewIndexSegmentsResult;
}

export class AlgoliaIndexSegmentManagerServiceImpl implements AlgoliaIndexSegmentManagerService {
    private static config = {
        apiKeyTTLHours: 8,
        /**
         * For example, if set to 30, API keys will stop functioning 30 minutes after they are removed from cache.
         */
        apiKeyTTLExpiryDiffMinutes: 30,
        maxKeysToCache: 5_000,
    };

    private apiKeysCache: Cache<string>;

    private get algolia() {
        return this.app.services.algolia;
    }

    constructor(private readonly app: FdrApplication) {
        this.apiKeysCache = new Cache(AlgoliaIndexSegmentManagerServiceImpl.config.maxKeysToCache, SECONDS_IN_ONE_HOUR);
    }

    public generateIndexSegmentsForDefinition({
        dbDocsDefinition,
        url,
    }: {
        dbDocsDefinition: DocsV1Db.DocsDefinitionDb;
        url: string;
    }): GenerateNewIndexSegmentsResult {
        const navigationConfig = dbDocsDefinition.config.navigation;

        return visitDbNavigationConfig<GenerateNewIndexSegmentsResult>(navigationConfig, {
            versioned: (config) => {
                const configSegmentTuples = config.versions.map((v) => {
                    const indexSegment = this.generateNewIndexSegmentForUnversionedNavigationConfig({
                        url,
                        version: { id: v.version, urlSlug: v.urlSlug },
                    });
                    return [v.config, indexSegment] as const;
                });
                return {
                    type: "versioned",
                    configSegmentTuples,
                };
            },
            unversioned: (config) => {
                const indexSegment = this.generateNewIndexSegmentForUnversionedNavigationConfig({
                    url,
                });
                return {
                    type: "unversioned",
                    configSegmentTuple: [config, indexSegment] as const,
                };
            },
        });
    }

    private generateNewIndexSegmentForUnversionedNavigationConfig({
        version,
        url,
    }: {
        version?: DocsVersion;
        url: string;
    }): IndexSegment {
        const indexSegmentId = this.generateUniqueIdForIndexSegment({ url, version });
        const searchApiKey = this.generateAndCacheApiKey(indexSegmentId);
        return version != null
            ? {
                  type: "versioned",
                  id: indexSegmentId,
                  searchApiKey,
                  version,
              }
            : {
                  type: "unversioned",
                  id: indexSegmentId,
                  searchApiKey,
              };
    }

    private generateUniqueIdForIndexSegment({ version, url }: { version?: DocsVersion; url: string }) {
        const parts: string[] = ["seg", url];
        if (version != null) {
            parts.push(kebabCase(version.id));
        }
        parts.push(uuidv4());
        return parts.join("_");
    }

    public getOrGenerateSearchApiKeyForIndexSegment(indexSegmentId: string) {
        const cachedKey = this.apiKeysCache.get(indexSegmentId);
        if (typeof cachedKey === "string") {
            return cachedKey;
        } else {
            return this.generateAndCacheApiKey(indexSegmentId);
        }
    }

    public getSearchApiKeyForIndexSegment(indexSegmentId: string) {
        const cachedKey = this.apiKeysCache.get(indexSegmentId);
        return cachedKey;
    }

    public generateAndCacheApiKey(indexSegmentId: string) {
        const now = new Date();
        const cacheUntil = addHours(now, AlgoliaIndexSegmentManagerServiceImpl.config.apiKeyTTLHours);
        const validUntil = addMinutes(
            cacheUntil,
            AlgoliaIndexSegmentManagerServiceImpl.config.apiKeyTTLExpiryDiffMinutes,
        );
        const key = this.algolia.generateSearchApiKey(`indexSegmentId:${indexSegmentId}`, validUntil);
        this.tryCacheApiKey(indexSegmentId, key);
        return key;
    }

    private tryCacheApiKey(indexSegmentId: string, apiKey: string) {
        try {
            this.apiKeysCache.set(indexSegmentId, apiKey);
        } catch {
            // Cache is full, ignore error
            return;
        }
    }
}
