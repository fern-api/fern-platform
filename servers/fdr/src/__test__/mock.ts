import { APIV1Db, DocsV1Db } from "@fern-api/fdr-sdk";
import { FdrApplication, type FdrConfig } from "../app";
import { type FdrServices } from "../app/FdrApplication";
import { ConfigSegmentTuple, type AlgoliaSearchRecord, type AlgoliaService } from "../services/algolia";
import { type AuthService } from "../services/auth";
import { OrgIdsResponse } from "../services/auth/AuthService";
import { RevalidatedPathsResponse, RevalidatorService } from "../services/revalidator/RevalidatorService";
import {
    FailedToDeleteIndexSegment,
    FailedToRegisterDocsNotification,
    FailedToRevalidatePathsNotification,
    GeneratingDocsNotification,
    SlackService,
} from "../services/slack/SlackService";
import { ParsedBaseUrl } from "../util/ParsedBaseUrl";

export class MockAlgoliaService implements AlgoliaService {
    generateSearchApiKey(_filters: string): string {
        return "";
    }

    async deleteIndexSegmentRecords(_indexSegmentIds: string[]): Promise<void> {
        return;
    }

    async generateSearchRecords(_: {
        docsDefinition: DocsV1Db.DocsDefinitionDb;
        apiDefinitionsById: Map<string, APIV1Db.DbApiDefinition>;
        configSegmentTuples: ConfigSegmentTuple[];
    }): Promise<AlgoliaSearchRecord[]> {
        return [];
    }

    async uploadSearchRecords(_records: AlgoliaSearchRecord[]): Promise<void> {
        return;
    }
}

class MockAuthService implements AuthService {
    orgIds: string[];

    constructor({ orgIds }: { orgIds: string[] }) {
        this.orgIds = orgIds;
    }

    async checkUserBelongsToOrg(): Promise<void> {
        return;
    }

    async getOrgIdsFromAuthHeader(_authHeader: { authHeader: string | undefined }): Promise<OrgIdsResponse> {
        return {
            type: "success",
            orgIds: new Set<string>(this.orgIds),
        };
    }

    checkOrgHasSnippetsApiAccess({
        authHeader,
        orgId,
        failHard,
    }: {
        authHeader: string | undefined;
        orgId: string;
        failHard?: boolean | undefined;
    }): Promise<boolean> {
        return Promise.resolve(false);
    }

    checkOrgHasSnippetTemplateAccess({
        authHeader,
        orgId,
        failHard,
    }: {
        authHeader: string | undefined;
        orgId: string;
        failHard?: boolean | undefined;
    }): Promise<boolean> {
        return Promise.resolve(false);
    }

    async getWorkOSOrganization(_orgId: { orgId: string }): Promise<string | undefined> {
        return undefined;
    }
}

class MockSlackService implements SlackService {
    async notify(_message: string, _err: unknown): Promise<void> {
        return;
    }

    async notifyFailedToRegisterDocs(_request: FailedToRegisterDocsNotification): Promise<void> {
        return;
    }

    async notifyFailedToRevalidatePaths(_request: FailedToRevalidatePathsNotification): Promise<void> {
        return;
    }

    async notifyFailedToDeleteIndexSegment(_request: FailedToDeleteIndexSegment): Promise<void> {
        return;
    }

    async notifyGeneratedDocs(_request: GeneratingDocsNotification): Promise<void> {
        return;
    }
}

class MockRevalidatorService implements RevalidatorService {
    async revalidate(_params: { baseUrl: ParsedBaseUrl }): Promise<RevalidatedPathsResponse> {
        return {
            successful: [],
            failed: [],
            revalidationFailed: false,
        };
    }
}

export const baseMockFdrConfig: FdrConfig = {
    awsAccessKey: "",
    awsSecretKey: "",
    publicDocsS3: {
        bucketName: "fdr",
        bucketRegion: "us-east-1",
        urlOverride: "http://s3-mock:9090",
    },
    privateDocsS3: {
        bucketName: "fdr",
        bucketRegion: "us-east-1",
        urlOverride: "http://s3-mock:9090",
    },
    privateApiDefinitionSourceS3: {
        bucketName: "fdr",
        bucketRegion: "us-east-1",
        urlOverride: "http://s3-mock:9090",
    },
    venusUrl: "",
    domainSuffix: ".docs.buildwithfern.com",
    algoliaAppId: "",
    algoliaAdminApiKey: "",
    algoliaSearchIndex: "",
    algoliaSearchApiKey: "",
    algoliaSearchV2Domains: [],
    slackToken: "",
    logLevel: "debug",
    docsCacheEndpoint: process.env["DOCS_CACHE_ENDPOINT"] || "",
    enableCustomerNotifications: false,
    applicationEnvironment: "mock",
    redisEnabled: false,
    redisClusteringEnabled: false,
};

export function getMockFdrConfig(overrides?: Partial<FdrConfig>): FdrConfig {
    if (overrides) {
        return {
            ...baseMockFdrConfig,
            ...overrides,
        };
    }
    return baseMockFdrConfig;
}

export function createMockFdrApplication({
    orgIds,
    services,
    configOverrides,
}: {
    orgIds?: string[];
    services?: Partial<FdrServices>;
    configOverrides?: Partial<FdrConfig>;
}) {
    return new FdrApplication(getMockFdrConfig(configOverrides), {
        auth: new MockAuthService({
            orgIds: orgIds ?? [],
        }),
        algolia: new MockAlgoliaService(),
        slack: new MockSlackService(),
        revalidator: new MockRevalidatorService(),
        ...services,
    });
}
