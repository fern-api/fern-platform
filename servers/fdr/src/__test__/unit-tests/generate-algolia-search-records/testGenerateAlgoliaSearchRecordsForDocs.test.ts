import {
    APIV1Write,
    DocsV1Write,
    FdrAPI,
    SDKSnippetHolder,
    convertAPIDefinitionToDb,
    convertDocsDefinitionToDb,
    visitDbNavigationConfig,
} from "@fern-api/fdr-sdk";
import { resolve } from "path";
import type { AlgoliaSearchRecord } from "../../../services/algolia";
import { AlgoliaIndexSegmentManagerServiceImpl } from "../../../services/algolia-index-segment-manager";
import { AlgoliaSearchRecordGeneratorV2 } from "../../../services/algolia/AlgoliaSearchRecordGeneratorV2";
import { createMockFdrApplication } from "../../mock";

const FIXTURES_DIR = resolve(__dirname, "fixtures");
const FIXTURES: Fixture[] = [
    {
        name: "primer",
    },
    {
        name: "vellum",
    },
    {
        name: "candid",
    },
    {
        name: "humanloop",
    },
];

function loadDocsDefinition(fixture: Fixture) {
    const filePath = resolve(FIXTURES_DIR, fixture.name, "docs.json");
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return require(filePath) as DocsV1Write.DocsDefinition;
}

function loadApiDefinition(fixture: Fixture, id: string) {
    const filePath = resolve(FIXTURES_DIR, fixture.name, "apis", `${id}.json`);
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return require(filePath) as APIV1Write.ApiDefinition;
}

type FilteredSearchRecord = Omit<AlgoliaSearchRecord, "objectID">;

function filterSearchRecord(record: AlgoliaSearchRecord): FilteredSearchRecord {
    const { objectID: _, ...rest } = record;
    return rest;
}

interface Fixture {
    name: string;
    only?: boolean;
}

const EMPTY_SNIPPET_HOLDER = new SDKSnippetHolder({
    snippetsBySdkId: {},
    snippetsConfigWithSdkId: {},
    snippetTemplatesByEndpoint: {},
    snippetsBySdkIdAndEndpointId: {},
    snippetTemplatesByEndpointId: {},
});

describe("generateAlgoliaSearchRecordsForDocs", () => {
    for (const fixture of FIXTURES) {
        const { only = false } = fixture;
        (only ? it.only : it)(
            `${JSON.stringify(fixture)}`,
            async () => {
                const docsDefinition = convertDocsDefinitionToDb({
                    writeShape: loadDocsDefinition(fixture),
                    files: {},
                });

                const preloadApiDefinitions = () => {
                    const apiIdDefinitionTuples = docsDefinition.referencedApis.map((id) => {
                        const apiDef = loadApiDefinition(fixture, id);
                        return [id, convertAPIDefinitionToDb(apiDef, id, EMPTY_SNIPPET_HOLDER)] as const;
                    });

                    return Object.fromEntries(apiIdDefinitionTuples);
                };

                const apiDefinitionsById = preloadApiDefinitions();
                const recordsWithoutIds: Omit<AlgoliaSearchRecord, "objectID">[] = [];
                const navigationConfig = docsDefinition.config.navigation;
                const generator = new AlgoliaSearchRecordGeneratorV2({ docsDefinition, apiDefinitionsById });

                if (navigationConfig == null) {
                    throw new Error("Navigation config is required for this test");
                }

                visitDbNavigationConfig(navigationConfig, {
                    versioned: (config) => {
                        config.versions.forEach((v) => {
                            const indexSegmentRecords = generator.generateAlgoliaSearchRecordsForSpecificDocsVersion(
                                v.config,
                                {
                                    type: "versioned",
                                    id: FdrAPI.IndexSegmentId(`${v.version}-constant`),
                                    searchApiKey: "api_key",
                                    version: { id: v.version, urlSlug: v.urlSlug },
                                },
                            );
                            recordsWithoutIds.push(...indexSegmentRecords.map(filterSearchRecord));
                        });
                    },
                    unversioned: (config) => {
                        const records = generator.generateAlgoliaSearchRecordsForSpecificDocsVersion(config, {
                            type: "unversioned",
                            id: FdrAPI.IndexSegmentId("constant"),
                            searchApiKey: "api_key",
                        });
                        recordsWithoutIds.push(...records.map(filterSearchRecord));
                    },
                });

                expect(recordsWithoutIds).toMatchSnapshot();
            },
            90_000,
        );
    }
});

describe("generateIndexSegmentsForDefinition", () => {
    const indexSegmentManager = new AlgoliaIndexSegmentManagerServiceImpl(createMockFdrApplication({}));
    it("should strip spaces from the version ID", () => {
        const segments = indexSegmentManager.generateIndexSegmentsForDefinition({
            dbDocsDefinition: {
                type: "v3",
                pages: {},
                referencedApis: [],
                files: {},
                config: {
                    navigation: {
                        versions: [
                            {
                                version: FdrAPI.VersionId("version with spaces"),
                                config: {
                                    items: [],
                                    landingPage: undefined,
                                },
                                urlSlug: undefined,
                                availability: undefined,
                            },
                            {
                                version: FdrAPI.VersionId("version with (special) chars"),
                                config: {
                                    items: [],
                                    landingPage: undefined,
                                },
                                urlSlug: undefined,
                                availability: undefined,
                            },
                        ],
                    },
                    root: undefined,
                    title: undefined,
                    defaultLanguage: undefined,
                    announcement: undefined,
                    navbarLinks: undefined,
                    footerLinks: undefined,
                    logoHeight: undefined,
                    logoHref: undefined,
                    favicon: undefined,
                    metadata: undefined,
                    redirects: undefined,
                    backgroundImage: undefined,
                    colorsV3: undefined,
                    layout: undefined,
                    typographyV2: undefined,
                    analyticsConfig: undefined,
                    integrations: undefined,
                    css: undefined,
                    js: undefined,
                    logo: undefined,
                    logoV2: undefined,
                    colors: undefined,
                    colorsV2: undefined,
                    typography: undefined,
                },
                jsFiles: undefined,
            },
            url: "https://example.com",
        });

        expect(segments.type).toBe("versioned");
        if (segments.type === "versioned") {
            expect(segments.configSegmentTuples[0]?.[1].id).toContain("version-with-spaces");
            expect(segments.configSegmentTuples[1]?.[1].id).toContain("version-with-special-chars");
        }
    });
});
