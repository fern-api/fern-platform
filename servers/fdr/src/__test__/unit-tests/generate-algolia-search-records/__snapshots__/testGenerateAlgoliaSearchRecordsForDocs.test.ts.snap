version https://git-lfs.github.com/spec/v1
oid sha256:7179caa7e96fc0dbf96fead65062fd2f705aa297f08833843c4f986d463bcd48
size 308860
